spharpy/spherical.py

```diff
@@ -1,6 +1,431 @@
+        channel_convention : str, optional
+            Channel ordering convention, either 'acn' or 'fuma'. The default is 'acn'.
+            (FuMa is only supported up to 3rd order)
+        inv_transform_type : str, optional
+            Inverse transform type, either 'pseudo_inverse' or 'quadrature'. The default is None.
+        normalization : str, optional
+            Normalization convention, either 'n3d', 'maxN' or 'sn3d'. The default is 'n3d'.
+            (maxN is only supported up to 3rd order)
+        """
+        self.n_max = n_max
+        self.coords = coords  # coordinates : :doc:`pf.Coordinates <pyfar:classes/pyfar.coordinates>`
+        self.basis_type = basis_type
+        self.inv_transform_type = inv_transform_type
+        self.channel_convention = channel_convention  # channel ordering convention
+        self.normalization = normalization  # gain normalization convention
+        self.basis = None
+        self.basis_gradient_theta, self.basis_gradient_phi = (None, None)
+        self._basis_inv = None
+        self._basis_inv_gradient_theta, self._basis_inv_gradient_phi = (None, None)
+
+    def _validate_parameters(self):
+        if not isinstance(self.coords, Coordinates):
+            raise TypeError("coords must be a pyfar.Coordinates object")
+        assert self.n_max >= 0, "n_max must be a positive integer"
+        assert self.channel_convention in ['acn', 'fuma'], ("Invalid channel convention, currently only 'acn' "
+                                                            "and 'fuma' are supported")
+        assert self.normalization in ['n3d', 'maxN', 'sn3d'], ("Invalid normalization, "
+                                                               "currently only 'n3d', 'maxN', 'sn3d' are "
+                                                               "supported")
+        assert self.basis_type in ['complex', 'real'], ("Invalid basis type, currently only 'complex' and 'real' "
+                                                        "are supported")
+        assert self.inv_transform_type in ['pseudo_inverse', 'quadrature', None], (
+            "Invalid inverse transform type, "
+            "currently only 'pseudo_inverse' "
+            "and 'quadrature' are supported")
+
+    @lru_cache(maxsize=128)
+    def _compute_basis(self):
+        logger.info("Computing basis matrix for n_max=%d", self.n_max)
+        if self.basis_type == 'complex':
+            self.basis = spherical_harmonic_basis(self.n_max, self.coords,
+                                                  self.normalization, self.channel_convention)
+        elif self.basis_type == 'real':
+            self.basis = spherical_harmonic_basis_real(self.n_max, self.coords,
+                                                       self.normalization, self.channel_convention)
+        else:
+            raise ValueError("Invalid signal type, should be either 'complex' or 'real'")
+
+    def compute_basis(self):
```

Comment:
Maybe call this only `basis`, since the user does not care if it is actually computed each time this is called? In any case, it should have a docstring.

Reply:
agreed

Reply:
Should be removed, right?

Reply:
You mean the comment or the compute_* properties?

Reply:
turn compute_basis into a private function (and compute_gradietn, etc)